### PR TITLE
Extend DateTimeFormatCheck to work for multiseries

### LINKED ISSUE #4301: Extend DataTimeFormatCheck data check to support multiseries

### PR DESCRIPTION
**Pull request by MichaelFu512**
### Pull Request Description
Extend DateTimeFormatCheck to work for multiseries, works on stacked datasets.

Closes #4301

### Small example of what this does
Say we had a multiseries problem with a stacked dataset, with two series (0, 1) and goes from 2021-01-01 to 2021-01-15.
Screen Shot 2023-09-07 at 2 05 33 PM

For Series 1, the date "2021-01-07" is missing (line 14) while Series 0 has an extra date "2021-01-29" (line 31).

When we use the datacheck and run validate:
Screen Shot 2023-09-07 at 2 09 40 PM

It gives the following result:
Screen Shot 2023-09-07 at 2 11 32 PM

### CODE REVIEW

**Comment by machineFL** [comment]
No linked issues found. Please add the corresponding issues in the pull request description. <br/> [Use GitHub automation to close the issue when a PR is merged](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

File: evalml/data_checks/datetime_format_data_check.py [hunk 1]
```diff
@@ -199,52 +210,6 @@ def validate(self, X, y):
             ...     }
             ... ]
 
-            The column "Weeks" has a date that does not follow the weekly pattern, which is considered misaligned.
```
Comment (MichaelFu512) [inline evalml/data_checks/datetime_format_data_check.py:202]: Removed because it looks like a copy of line 156-200

File: evalml/data_checks/datetime_format_data_check.py [hunk 2]
```diff
@@ -378,110 +348,147 @@ def validate(self, X, y):
...
+                    ].reset_index(drop=True)
+                else:
+                    datetime_values = X[X[self.series_id] == series].index
+
+            # Check if the data is monotonically increasing
+            no_nan_datetime_values = datetime_values.dropna()
+            if not pd.DatetimeIndex(no_nan_datetime_values).is_monotonic_increasing:
+                messages.append(
+                    DataCheckError(
+                        message="Datetime values must be sorted in ascending order.",
```
Comment (MichaelFu512) [inline evalml/data_checks/datetime_format_data_check.py:369]: Unsure if I need to change the message here to say something like "Datetime values {at series x} must be sorted in ascending order" or if it's alright to just keep it general.

File: evalml/data_checks/datetime_format_data_check.py [hunk 3]
```diff
@@ -378,110 +402,147 @@ def validate(self, X, y):
...
-                ).to_dict(),
+        for series in series_datetime:
+            # if multiseries only select the datetimes corresponding to one series
+            if is_multiseries:
+                if self.datetime_column != "index":
+                    datetime_values = X[X[self.series_id] == series][
+                        self.datetime_column
+                    ].reset_index(drop=True)
+                else:
+                    datetime_values = X[X[self.series_id] == series].index
```
Comment (eccabay) [inline evalml/data_checks/datetime_format_data_check.py:416]: Mega nit: can you pull out `X[X[self.series_id] == series]` to before the inner conditional statement? No real reason other than readability.

File: evalml/data_checks/datetime_format_data_check.py [hunk 4]
```diff
@@ -378,110 +402,147 @@ def validate(self, X, y):
...
-                DataCheckError(
-                    message=f"Column '{col_name}' has datetime values missing between start and end date.",
-                    data_check_name=self.name,
-                    message_code=DataCheckMessageCode.DATETIME_IS_MISSING_VALUES,
-                ).to_dict(),
+            inferred_freq = ww_payload[0]
+            debug_object = ww_payload[1]
+            if inferred_freq is not None:
+                if is_multiseries:
+                    continue
```
Comment (eccabay) [inline evalml/data_checks/datetime_format_data_check.py:444]: Small simplification: `if inferred_freq is not None and not is_multiseries:`

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py [hunk 5]
```diff
@@ -349,35 +352,6 @@ def test_datetime_format_data_check_multiple_errors():
...
-            message_code=DataCheckMessageCode.DATETIME_IS_MISSING_VALUES,
-        ).to_dict(),
-        DataCheckError(
-            message="Column 'dates' has datetime values that do not align with the inferred frequency.",
-            data_check_name=datetime_format_check_name,
-            message_code=DataCheckMessageCode.DATETIME_HAS_MISALIGNED_VALUES,
-        ).to_dict(),
-        get_uneven_error("dates", ww_payload),
-    ]
-
```
Comment (eccabay) [inline evalml/tests/data_checks_tests/test_datetime_format_data_check.py:380]: Why was this deleted?

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py [hunk 6]
```diff
@@ -541,3 +515,287 @@ def test_datetime_many_duplicates_and_nans():
...
     assert result[2]["code"] == "DATETIME_NO_FREQUENCY_INFERRED"
+
+
+@pytest.mark.parametrize("input_type", ["pd", "ww"])
+@pytest.mark.parametrize(
+    "issue",
+    ["redundant", "missing", "uneven", "type_errors", None],
+)
+@pytest.mark.parametrize("datetime_loc", [1, "X_index"])
+def test_datetime_format_data_check_typeerror_uneven_intervals_multiseries(
```
Comment (eccabay) [inline evalml/tests/data_checks_tests/test_datetime_format_data_check.py:526]: This is a long test with a lot of repeated code with the single series version of this test. Could you combine this with the pre-existing one and parametrize? You should be able to abstract out the `repeat` calls to generate multiple series by using `pd.Series.repeat(1)` for the single series case.

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py [hunk 7]
```diff
@@ -541,3 +515,287 @@ def test_datetime_many_duplicates_and_nans():
...
+    )
+    with pytest.raises(
+        ValueError,
+        match="""series_id "not_series_id" is not in the dataset.""",
+    ):
+        datetime_format_check.validate(X, y)
+
+
+@pytest.mark.parametrize("n_missing", [2, 5, 7])
+def test_datetime_format_data_check_multiple_missing_multiseries(n_missing):
```
Comment (eccabay) [inline evalml/tests/data_checks_tests/test_datetime_format_data_check.py:674]: Same with this test - can it be combined with the single series version?

**Review by eccabay — CHANGES_REQUESTED** [review]
Just a few nitpicks and some testing consolidation, but otherwise looks great! Thanks for doing this

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py (hunk 5 above)
Comment (MichaelFu512) [reply evalml/tests/data_checks_tests/test_datetime_format_data_check.py:380]: It looks like an exact copy of line 323 - 349

**Review by eccabay — APPROVED** [review]

**Review by jeremyliweishih — APPROVED** [review]
LGTM - just some punctuation requested.
